[Ready] Remote plugins

KrauseFx opened the pull request:
Working on local and remote actions using simple Ruby classes (no gems)

**Open points:**

- [x] Update plugins to be classes and not methods
- [x] `danger new_plugin` command
- [x] Tests

DangerCI commented:
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%">2 Errors</th>
     </tr>
  </thead>
  <tbody>
        <tr>
      <td>:no_entry_sign:</td>
      <td>Please provide a summary in the Pull Request description</td>
    </tr>
        <tr>
      <td>:no_entry_sign:</td>
      <td><p>Please include a CHANGELOG entry. 
You can find it at <a href="https://github.com/danger/danger/blob/master/CHANGELOG.md">CHANGELOG.md</a>.</p>
</td>
    </tr>
    
  </tbody>
</table>
  

  
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%">2 Warnings</th>
     </tr>
  </thead>
  <tbody>
        <tr>
      <td>:warning:</td>
      <td>Pull Request is Work in Progress</td>
    </tr>
        <tr>
      <td>:warning:</td>
      <td>Tests were not updated</td>
    </tr>
    
  </tbody>
</table>
  

  


<p align="right" data-meta="generated_by_danger" data-base-commit="" data-head-commit="" >
  Generated by :no_entry_sign: <a href="https://github.com/danger/danger/">danger</a>
</p>

KrauseFx commented:
This PR is ready for review. Circle is failing for a random bundler reason. 
I don't know why Travis fails with 

```
lib/assets/ActionTemplate.rb:4:28: E: unexpected token tLT
(Using Ruby 2.0 parser; configure using TargetRubyVersion parameter, under AllCops)
      class [[CLASS_NAME]] < Plugin
                           ^
lib/assets/ActionTemplate.rb:20:1: E: unexpected token kEND
(Using Ruby 2.0 parser; configure using TargetRubyVersion parameter, under AllCops)
end
^^^
44 files inspected, 2 offenses detected
RuboCop failed!
```

This file is excluded in `.rubocop.yml` and works locally. Any ideas?

orta commented:
@dbgrandi - I'll let you review etc 

I don't think `danger new_plugin` should exist though, it should be replaced by your `danger plugins create` is that close yet?

dbgrandi commented on CHANGELOG.md:
```diff
@@ -1,6 +1,12 @@
+* Added support for remote plugins
+* Added new `danger new_plugin` command to create plugins in the fastlane
 * Added new `modified_files.include?("rakelib/*_stats.rake") file globbing support - KrauseFx
+* Added printing of table summaries after running danger
```
You left the backtick around `modified_files.include?` on the previous line open, so the following lines may show up as monospaced.

KrauseFx replied:
Nice catch :+1:

dbgrandi commented on lib/assets/ActionTemplate.rb:
```diff
@@ -0,0 +1,20 @@
+      end
+    end
+  end
+end
```
Should this be `ActionTemplate.rb.template` or something to emphasize that it's not just an `.rb` file?

jeroenvisser101 replied:
or even rendering with erb?

jeroenvisser101 replied:
(that might be a bit overkill though...)

KrauseFx replied:
Yah, it's too much since we just replace one piece of the string

jeroenvisser101 replied:
yep, I thought the same after I actually checked that code :+1:

jeroenvisser101 replied:
however, the .tpl or .template would make it more obvious that it's a template. Also, since 'assets' is only templates, we could consider renaming that folder all together.

dbgrandi commented on lib/danger/dangerfile_dsl.rb:
```diff
@@ -58,18 +91,38 @@ def message(message)
+        # Plugins
+        class_name = method_sym.to_s.danger_class
+        if Danger::Dangerfile::DSL.const_defined?(class_name)
+          plugin_ref = Danger::Dangerfile::DSL.const_get(class_name)
```
Is it worth adding a check here to make sure the `plugin_ref` extends from `Plugin` and not just a freestanding method or something?

KrauseFx replied:
Yeah, great point, I'll add this! Also, I should add a `.run` method to the super class that raises an exception if you didn't implement it

dbgrandi commented on lib/danger.rb:
```diff
@@ -3,6 +3,7 @@
 require "danger/environment_manager"
 require "danger/commands/runner"
 require "danger/available_values"
+require "danger/core_ext/string"
```
Are these in a particular order? The ones below are all :abc:

KrauseFx replied:
That shouldn't really matter

dbgrandi commented:
Looks good! :telescope: 

I've still got a bunch to do on refactoring for the plugin manager.

orta commented on README.md:
```diff
@@ -102,6 +102,32 @@ snapshots_url = build_log.match(%r{https://eigen-ci.s3.amazonaws.com/\d+/index.h
+
+This will generate a new Ruby file which you can modify to fit your needs. 
+
+If you think your new plugin is useful for other developers too, feel free to submit a PR to danger by copying the action to `lib/danger/plugins`. 
```
Blocker; I don't want to see Danger getting filled with Actions the way fastlane does - people should be using gems for this stuff, not relying on us to keep everything working in one repo. Please remove this line.

jeroenvisser101 replied:
We could have one 'gem', `danger-useful` or similar, containing a bunch of useful (WIP, Not a contributor, No spec update) plugins that you could start with?

KrauseFx replied:
@jeroenvisser101 Great idea, this would be a separate repo in the `danger` org, maybe `danger-essentials`? 
We should make it easy for new users to get started with `danger` fast

jeroenvisser101 replied:
:+1: essentials makes total sense!

jeroenvisser101 replied:
@KrauseFx – How would we add multiple plugin classes to one plugin?

orta replied:
> We could have one 'gem', `danger-useful`

Terminology wise, I think `danger-core` would work here. If it's essentials, it should ship with danger. There's no point in over-abstracting too far yet.

The reason I call this a blocker is that I can see how much time & effort it takes felix to maintain fastlane, and that's got a big audience. Danger could have an even bigger audience, given it's usefulness in most dev communities. So, we want to make it _really, really_ easy for others to build on top of danger, and not _into_ danger. Specifically, so that we're not spending time on minutae of actions instead of building a better platform.

If it makes sense to use something in every danger file ( like `protect_files` does ) then it should be in with Danger, if we feel like it would be better for all of our provided actions to go into `danger-core` cool, but it needs to be all or nothing, and do we need that extra level of abstraction yet?

jeroenvisser101 replied:
We could limit `danger-{core,useful,essentials}` in scope by including guidelines in both the PR/Issue templates and `CONTRIBUTING.md` so people know what will be merged and what will have to be in a separate plugin/repo. Also, I think the help of having a very limited, but often-used plugins might help so much people get started with Danger.

> If it's essentials, it should ship with danger. There's no point in over-abstracting too far yet.

I do agree that this looks like over-abstracting for now, and we should hit that roadblock first before going around it.

orta replied:
I'm cool with making that jump at some point, but I want everything from `message` to `protect_files` to be in it, so that the actual "core/essentials" api is just another plugin ( that is included by default )

Agreed re: templates, once I get back from siberia I'll have a think

KrauseFx replied:
Alright, so basically we do leave the most important methods in `danger`, however we remove the comment about submitting PRs when this plugin is useful? 

If that's ok, I'll update the PR according to all the comments

KrauseFx commented:
Okay, addressed all the feedback and updated the PR, thanks everyone.

@orta Do you want `danger new_plugin` to be changed to `danger plugins create`? If so, could you point me to the right direction on how to implement this?

KrauseFx commented:
@dbgrandi @jeroenvisser101 I'd appreciated another review of the changes :+1:

jeroenvisser101 commented:
@KrauseFx LGTM! :rocket:

orta commented:
> Do you want danger new_plugin to be changed to danger plugins create

No, we'll have a real one coming in one of these days with #131, I'd recommend dropping it from this PR

KrauseFx commented:
@orta I should completely remove the `danger new_plugin` command? How are people supposed to create new local plugins? This is different from #131 as it's mostly used for local additions to danger

KrauseFx commented:
@orta just wanted to ask you about the last remaining open point https://github.com/danger/danger/pull/156#issuecomment-204472675

orta commented:
Keep it in, but don't be surprised if it gets removed once we can push gems properly

KrauseFx commented:
Alright, resolved merge conflicts - working on finishing things up now :+1:

KrauseFx commented:
I pushed a few more commits with more tests and documentation, they should be live soon, just waiting for https://twitter.com/githubstatus/status/717464295454470145.

Ready to merge from my side. Tests are passing :+1:

orta commented:
Let's do it! :+1:

KrauseFx commented:
Oh yeah, thanks for merging @orta :+1:
